### PR TITLE
fix(utils): modify atomWithHash to keep pathname when the replaceState option is set to true

### LINKED ISSUE #1110: [atomWithHash] replaceState should probably keep current path?
Setting `options.replaceState = true` causes the atom to overwrite current path with `#{jotai_hash}`. I would think what it should instead write `previous/path#{jotai_hash}`. Current implementation does not work well in SPAs that use `react-router` or similar libraries.

If this is the case I could write a PR to solve this.

### PR DESCRIPTION
**Pull request by jakkku**
## Related Issues

Fixes #1110 .

## Summary

If the replaceState option is set to true, the pathname is overwritten with a hash when returning to the page using `history.back`.

An error occurred in the codesandbox environment, but no error occurred in the test code environment.
But I added a test case.

This is my first time working in Jotai's codebase. Please let me know if there are any additional considerations

## Check List

- [x] `yarn run prettier` for formatting code and docs

### CODE REVIEW

File: tests/utils/atomWithStorage.test.tsx [hunk 1]
```diff
@@ -470,4 +471,56 @@ describe('atomWithHash', () => {
     await findByText('visible')
     expect(window.location.hash).toEqual('')
   })
+
```
Comment (dai-shi) [inline tests/utils/atomWithStorage.test.tsx:474]:
```suggestion
})

describe('atomWithHash with react-router-dom', () => {
```

**Review by dai-shi — APPROVED** [review]
LGTM

**Comment by dai-shi** [comment]
Works good: https://codesandbox.io/s/react-typescript-forked-kub7ji

Please check CI errors.

**Comment by jakkku** [comment]
@dai-shi  When I check the CI errors, errors occur in `yarn install --frozen-lockfile`.
I added react-router-dom library to write test code, this seems to be the cause.

Would it be a good idea to modify the test code in a way that doesn't use the react-router-dom library?

**Comment by dai-shi** [comment]
I think you just need to run `yarn` once.

**Comment by dai-shi** [comment]
> Would it be a good idea to modify the test code in a way that doesn't use the react-router-dom library?

If it's reasonably possible, it'd be better anyways.

**Comment by dai-shi** [comment]
How hard is it to write the test without react-router-dom?

**Comment by jakkku** [comment]
> How hard is it to write the test without react-router-dom?

I think it is possible, so I will try

File: tests/utils/atomWithStorage.test.tsx [hunk 2]
```diff
@@ -471,3 +471,47 @@ describe('atomWithHash', () => {
     expect(window.location.hash).toEqual('')
   })
 })
+
+describe('atomWithHash with page move', () => {
```
Comment (dai-shi) [inline tests/utils/atomWithStorage.test.tsx:475]:
```suggestion

```

**Review by dai-shi — APPROVED** [review]
Nice. Now, I think we can have it in the same "describe".

**Pull request merged**